ddspipe_core/src/cpp/monitoring/producers/TopicsMonitorProducer.cpp

```diff
@@ -0,0 +1,333 @@
+    {
+        topic_data_[topic_name].name(topic_name);
+        topic_data_[topic_name].type_name(type_name);
+    }
+
+    if (!participant_data_.count(topic_name) || !participant_data_[topic_name].count(participant_id))
+    {
+        // First message received for topic & participant
+        // Save the participant_id
+        participant_data_[topic_name][participant_id].participant_id(participant_id);
+    }
+
+    // Increase the count of the received messages
+    participant_data_[topic_name][participant_id].msgs_received(participant_data_[topic_name][participant_id].msgs_received() +
+            1);
+}
+
+void TopicsMonitorProducer::msg_lost(
+        const types::DdsTopic& topic,
+        const types::ParticipantId& participant_id)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_lost.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    const auto& topic_name = topic.m_topic_name;
+    const auto& type_name = topic.type_name;
+
+    if (!participant_data_.count(topic_name) || !participant_data_[topic_name].count(participant_id))
+    {
+        // First message lost for topic & participant
+        // Save the participant_id
+        participant_data_[topic_name][participant_id].participant_id(participant_id);
+    }
+
+    // Increase the count of the lost messages
+    participant_data_[topic_name][participant_id].msgs_lost(participant_data_[topic_name][participant_id].msgs_lost() + 1);
+}
+
+void TopicsMonitorProducer::type_discovered(
+        const std::string& topic_name,
+        const std::string& type_name)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_lost.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    if (!topic_data_.count(topic_name))
+    {
+        topic_data_[topic_name].name(topic_name);
+        topic_data_[topic_name].type_name(type_name);
+    }
+
+    topic_data_[topic_name].type_discovered(true);
+}
+
+void TopicsMonitorProducer::type_mismatch(
+        const types::DdsTopic& topic)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_lost.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    const auto& topic_name = topic.m_topic_name;
+    const auto& type_name = topic.type_name;
+
+    if (!topic_data_.count(topic_name))
+    {
+        topic_data_[topic_name].name(topic_name);
+        topic_data_[topic_name].type_name(type_name);
+    }
+
+    topic_data_[topic.m_topic_name].type_mismatch(true);
```

Comment:
I guess my previous comment would make this more difficult.

Reply:
Indeed. The thing is that in the case of `type_discovered` and `type_mismatch` it makes sense that the map's identifier is the `type_name`. But in the case of `msg_received` and `msg_lost`, it should be the `DdsTopic`.